grayscale.py

```diff
@@ -392,97 +398,88 @@ def brightness_sync(self, c):
-    @micropython.native
-    def _display_thread(self):
-        import utime        # sometimes utime global gets lost for this thread, so we reimport it here
-        buffer1 = self._buffer1
-        buffer2 = self._buffer2
-        buffer3 = self._buffer3
-        post_frame_adj1 = self.post_frame_adj[0]
-        post_frame_adj2 = self.post_frame_adj[1]
-        post_frame_adj3 = self.post_frame_adj[2]
-        state = self._state
-        display_thread_frame = self._display_thread_frame
-        state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Running
-        try:
-            while True:
-                while state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Running:
-                    display_thread_frame(0, post_frame_adj1, buffer1)
-                    display_thread_frame(1, post_frame_adj2, buffer2)
-                    display_thread_frame(2, post_frame_adj3, buffer3)
-                if state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Stopping:
-                    self._teardown_clear_render_buffer1()
-                    self.dc(1)
-                    self.spi.write(self._buffer1)
-                    state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Stopped
-                    return
-                if state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Pausing:
-                    state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Paused
-                    while state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Resuming:
-                        idle()
-                    state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Running
-        except Exception as e:
-            self._state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Exception
-            print(e)
+        b1:ptr32 = ptr32(self.buffer1) ; b2:ptr32 = ptr32(self.buffer2)
+        _b1:ptr32 = ptr32(self._buffer1) ; _b2:ptr32 = ptr32(self._buffer2) ; _b3:ptr32 = ptr32(self._buffer3)
 
+        fn:int ; i:int ; t0:int
+        v1:int ; v2:int ; contrast:int
 
-    @micropython.viper
-    def _teardown_clear_render_buffer1(self):
-        b1:ptr32 = ptr32(self._buffer1)
-        i:int = 0
-        while i < 90:
-            b1[i] = 0
-            i += 1
+        state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Running
+        while True:
+            while state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Running:
+                fn = 0
+                while fn < 3:
+                    t0 = ticks_us()
+                    dc(0)
+                    spi_write(pre_frame_cmds)
+                    dc(1)
+                    spi_write(buffers[fn])
+                    dc(0)
+                    spi_write(post_frame_adj[fn])
+                    sleep_us(Grayscale_pre_frame_time_us - int(ticks_diff(ticks_us(), t0)))
+                    t0 = ticks_us()
+                    spi_write(post_frame_cmds)
+                    spi_write(post_frame_adj[fn])
```

Comment:
Why do you send the contrast twice?

Reply:
I thought it was needed to avoid a glitch (maybe banding? can't remember for sure), but it looks like the first one isn't actually needed. I'll remove it.

Reply:
Scratch that. After removing it I found that it _does_ result in a glitch (bottom of the screen flashes a different colour). I've left it in.